jobs: prevent parallel execution and plugin specific scheduler

Exirel opened the pull request:
Closes #1144 

### Description

Initially, it was only about preventing parallel execution of jobs, but then I realized I needed way more than that:

* changed the job object to be registered only once for any set of intervals,
* changed the job object to have similar meta-data than plugin's rules
* added a new scheduler, that can be used in a plugin-specific context (similar to the plugin's rules manager)

Also, I realized that the "catch up" mechanism makes no sense: either the job can be run in time, or it can't, but it won't try to run it more than once because it's late. I can see why it was like that before, but there are several problems with how it was implemented:

* the scheduler is not reliable: it ticks at best every second, at worst every now and then
* a job isn't aware of the catch up mechanism: is it late? is it on time? it can't know, so it can't adapt itself

In that regard, it's best not to over-engineer the solution, and to just take the most optimistic and safe approach: execute only once at a time, then see if it should execute again or a bit later, given the configured intervals.

And here we are, with a lot of changes for something that seem easy at first. The good news is that it was necessary anyway to work on a better documentation spec for Sopel: in the future, we'll be able to list not only rules and commands, but also jobs! But that will be for a future PR.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

dgw commented on sopel/bot.py:
```diff
@@ -394,16 +398,17 @@ def get_plugin_meta(self, name):
+
+        This method was used to unregister anything (rules, commands, urls,
+        jobs, and shutdown methods), but since everything can be done by other
+        mean, there is no use for it anymore.
```
```suggestion
        means, there is no use for it anymore.
```

Typo fixing reflex.

dgw commented on sopel/bot.py:
```diff
@@ -476,20 +481,8 @@ def register_callables(self, callables):
-    def unregister_jobs(self, jobs):
-        for job in jobs:
-            callable_name = getattr(job, "__name__", 'UNKNOWN')
-            self.scheduler.remove_callable_job(job)
-            LOGGER.debug('Job callable removed: %s', callable_name)
+            job = sopel.tools.jobs.Job.from_callable(self.settings, func)
+            self._scheduler.register(job)
```
Seems to me we should keep either both of these or neither, but not only one.

dgw commented on sopel/plugins/jobs.py:
```diff
@@ -0,0 +1,91 @@
+
+
+class Scheduler(jobs.JobScheduler):
+    """Scheduler of plugin jobs.
```
Should document the params to `__init__()` somewhere.

dgw commented on sopel/plugins/jobs.py:
```diff
@@ -0,0 +1,91 @@
+        LOGGER.debug('Successfully unregistered all jobs')
+
+    def remove_callable_job(self, callable):
+        """Remove ``callable`` from the job queue."""
```
```suggestion
        """Remove ``callable`` from the job queue.

        :param callable: the callable to remove
        :type callable: :term:`callable`
        """
```

It's kinda silly, but consistency is nice. And our new docstring guidelines (#1794) demand param docs.

Exirel replied:
Sadly, `callable` is not a term of the Python glossary. I'll use `function` instead.

Exirel replied:
(docstring updated on the base class)

dgw commented on sopel/tools/jobs.py:
```diff
@@ -39,10 +39,11 @@ def __init__(self, manager):
-    def add_job(self, job):
-        """Add a Job to the current job queue."""
+    def register(self, job):
+        """Regsister a Job to the current job queue."""
```
```suggestion
        """Register a Job to the current job queue."""
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -139,19 +142,104 @@ class Job(object):
     next so it should only be called right after the function was called.
     """
+    @classmethod
+    def kwargs_from_callable(cls, settings, handler):
```
I'm not going to attach a note to every single one, but so many new methods here are lacking docstrings. :/

Exirel replied:
Yup, next commit will be full of them. I didn't want to add docstrings at first, because it's all internal. But heh. Eventually, I prefer *with* docstrings!

dgw commented on sopel/tools/jobs.py:
```diff
@@ -161,41 +249,43 @@ def is_ready_to_run(self, at_time):
+        )
 
-    def next(self):
+    def next(self, current_time):
```
Update docstring with added param.

dgw commented on sopel/tools/jobs.py:
```diff
@@ -161,41 +249,43 @@ def is_ready_to_run(self, at_time):
+        """Execute the job's handler and mark itself as running.
 
-        Example result::
+        :param manager: used as argument to the job's handler
```
`:type manager:`?

Exirel replied:
Mixed.

Exirel replied:
Ended with `object`. That will do.

dgw commented on sopel/bot.py:
```diff
@@ -394,16 +398,17 @@ def get_plugin_meta(self, name):
         version='7.1',
         removed_in='8.0')
     def unregister(self, obj):
```
In hindsight, I'm not sure removing functionality from this method bit by bit was the right approach. We're left with a "deprecated" method that actually won't do what it used to any more, so it's no longer backward compatible.

This is partly on me for not thinking of it before, but since this is kind of a bot internal… should we just delete this now and skip the deprecation cycle? It's either that or restore all functionality to work just like it did in 7.0.4 (and earlier), and maintain it up until 8.0.

Exirel replied:
Hm yeah we should remove it entirely. I didn't do it at first because... I don't know? Maybe I removed them bit by bit in so many different PRs that I didn't think about it too much.

Exirel commented:
OK, there is more to say about the register/unregister situation in the bot.

1. `bot.register` expects 4 parameters (callables, jobs, urls, shutdown), but `bot.unregister` expect just one callable and deal with it. What is that??? I think we should remove both, and I'd argue that we should not bother with deprecation cycle in this case.
2. `bot.register_*`  methods make sense, because they allow to register a callable, that will be transformed into some sort of object, tied to a plugin.
3. `bot.unregister_*`  methods don't, because what we usually want is to unregister a whole plugin, not just a single function (there should be other means to do that properly, like using rule's label, or command's name)
4. In fact, these `(un)register_*` methods could (should?) be private, because the *right* interface should be about adding/removing a plugin, and not dealing with individual parts. I remember when working on #1479 I had so much trouble making sense of the bot's interface approaching callables, jobs, urls, and so on
5. At some point, shutdown methods shouldn't be registered - the plugin handler can do that for us
6. Also the URL callback registration should be designed like the Rules Manager and the plugin's Scheduler, in a somewhat unified interface

It's a long process. So in this PR, I'll revert my changes on `unregister`, change what I did on `unregister_jobs`, and I'll do another PR for everything else mentioned in this comment, if that sounds good for you @dgw ?

Exirel commented:
@dgw also, I'll make document `sopel.tools.jobs`. I'll add a big `important` block to tell it's all internal an stuff, but if I don't do that, I'll have trouble documenting `sopel.plugins.jobs`.

Exirel commented:
@dgw so, you were totally right in your comments. Because I added *so much* docstrings, and because there was so much to fix/change/revert, I believe a "rebase" couldn't hurt in that specific case. Most of the docstrings have been updated, added, moved, or replaced, so I think you will have to re-review everything *anyway*.

As said earlier, I won't update the `bot.unregister` method again in this PR, I'll do a separate one for that, I think it's better this way. Also, I'm sorry for all the mess. I'm glad you are here doing the review, because it really makes Sopel better. Sometimes, I'm lazy, or I forget something, or I make an honest mistake. And usually, you're here to make sure it's fixed properly, and that's invaluable! :heart: 

So good luck with the review, hope it won't be too difficult this time.

dgw commented on sopel/plugins/jobs.py:
```diff
@@ -0,0 +1,111 @@
+
+        This class is a specific implementation of the scheduler, made to store
+        jobs by their plugins and be used by the bot (its ``manager``).
+        It follows a similar interface as the plugin's
+        :class:`rules manager <sopel.plugins.rules.Manager>`.
```
```suggestion
        It follows a similar interface as the
        :class:`plugin rules manager <sopel.plugins.rules.Manager>`.
```

That short-ass line tho, oh well.

dgw commented on sopel/plugins/jobs.py:
```diff
@@ -0,0 +1,111 @@
+    .. important::
+
+        This is an internal tool used by Sopel to manage its jobs. To register
+        a job, plugin author should uses :func:`sopel.module.interval`.
```
```suggestion
        a job, plugin authors should use :func:`sopel.module.interval`.
```

Let me guess… 🇫🇷? 😁

Exirel replied:
No, just me.

dgw commented on sopel/tools/jobs.py:
```diff
@@ -21,31 +23,74 @@
+        scheduler.start()  # run the thread forever
+
+    Then it runs forever until the :meth:`stop` method is called, usually when
+    the bot shutdowns.
```
```suggestion
    the bot shuts down.
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -21,31 +23,74 @@
         self.manager = manager
+        """Job manager, used as argument for jobs."""
         self.stopping = threading.Event()
+        """Stopping flag. See :meth:`stop`"""
```
```suggestion
        """Stopping flag. See :meth:`stop`."""
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -54,20 +99,45 @@ def stop(self):
+        """Run forever until :meth:`stop` is called.
+
+        This method waits at most a second between each iteration. At each step
+        it retrieves the jobs that are ready for execution, and, well, execute
+        them. See the :meth:`Job.execute` method for more information.
```
```suggestion
        it retrieves the jobs that are ready for execution, and executes them.
        See the :meth:`Job.execute` method for more information.
```

I do appreciate the bit of cheek, but don't think it belongs in this kind of documentation. :/

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+    :param intervals: set of intervals; each is a number of seconds between
+                      calls to ``handler``
+    :type intervals: :term:`iterable`
+    :param str plugin: optional job's plugin
+    :param str label: optional job's label
```
```suggestion
    :param str plugin: optional plugin name to which the job belongs
    :param str label: optional label (name) for the job
```

Further suggestions invited! I'm just doing my best with the slightly odd (I assume French-inspired) thing you sometimes do with possessives.

Exirel replied:
I need a (semi) advanced English grammar course. :-/

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+    :param str label: optional job's label
+    :param handler: function to be called when the job is ready to execute
+    :type handler: :term:`function`
+    :param str doc: option job's documentation
```
```suggestion
    :param str doc: optional documentation for the job
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+
+        if job.is_ready_to_run(time.time()):
+            job.execute(manager)  # marked as running
+            # "next times" have been updated, the job is not running
```
```suggestion
            # "next times" have been updated; the job is not running
```

Literally the hardest punctuation mark to learn how to use. 😹

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+    .. seealso::
+
+        The :class:`sopel.plugins.jobs.Scheduler` class is specifically
+        designed for plugin's jobs, expecting an instance of
```
```suggestion
        designed for plugins' jobs, expecting an instance of
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+    def kwargs_from_callable(cls, handler):
+        """Generate the keyword arguments to create a new instance.
+
+        :param callable handler: callable used to generate keyword arguments
```
I'm just curious how/if Sphinx renders the type `callable`, since below you split it into a separate `:type:` line with <tt>:term:\`function\`</tt>.

Exirel replied:
I think it doesn't. :-/

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+        # scheduling
+        now = time.time()
+        self.intervals = set(intervals)
+        """Set of intervals to execute the job at."""
```
```suggestion
        """Set of intervals at which to execute the job."""
```

"Don't end a sentence with a preposition," they say.

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+        self.is_running = threading.Event()
+        """Running flag: it tells if the job is running or not.
+
+        This flag is set and clear automatically by the :meth:`execute` method.
+        It is also set and clear when the job is used with the ``with``
+        statement::
```
```suggestion
        This flag is set and cleared automatically by the :meth:`execute`
        method. It is also set and cleared when the job is used with the
        ``with`` statement::
```

(I'm guessing at text wrapping width; too lazy to open my editor just for this.)

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+
+            with job:
+                # you do something before executing the job
+                # this ensure that the job is marked as "running"
```
```suggestion
                # this ensures that the job is marked as "running"
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+        .. note::
+
+            When set manually or with the ``with`` statement, the
+            :meth:`execute` won't clear this attribute.
```
```suggestion
            :meth:`execute` method won't clear this attribute itself.
```

Still make sense / match what you meant?

Exirel replied:
Actually it's even better.

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+        :return: the return value from the handler's execution
 
-            <Job(2013-06-14 11:01:36.884000, 20s, <function upper at 0x02386BF0>)>
+        This method execute the job's handler, and while doing so, it will
```
```suggestion
        This method executes the job's handler, and while doing so it will
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+
+        1. mark the job as running (set :attr:`is_running`) if not already set
+        2. execute the handler
+        3. if it wasn't already running before, it update the next time the
```
```suggestion
        3. if it wasn't already running before, update the next time the
```

dgw commented on sopel/tools/jobs.py:
```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+        .. note::
+
+            This method can be safely used with a ``with`` statement, as it
+            set/clear the :attr:`is_running` event only if not previously set
```
```suggestion
            sets/clears the :attr:`is_running` event only if not previously set
```

Exirel commented:
Change applied and fixed the docstring.

Exirel commented:
Done as well.
